File: baekjoon/사탕게임.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

int n;
string s[51];

void swap(int i, int j){
    
    char tmp = s[i][j];
    s[i][j] = s[i][j+1];
    s[i][j+1] = tmp;
}

void swap2(int i, int j){
    char tmp = s[i][j];
    s[i][j] = s[i+1][j];
    s[i+1][j] = tmp;
}
int check(){
    int ma = 0;
    int cnt = 1;
    for(int i=0; i<n; i++){
        cnt = 1;
        for(int j=0;j<n - 1;j++){
            if(s[i][j] == s[i][j+1]){
                cnt++;
                ma = max(cnt, ma);
            }
            else{
                cnt = 1;
            }
        }
    }
    
    for(int i=0; i<n; i++){
        cnt = 1;
        for(int j=0;j<n - 1;j++){
            if(s[j][i] == s[j+1][i]){
                cnt++;
                ma = max(cnt, ma);
            }
            else{
                cnt = 1;
            }
        }
    }
    ma = max(cnt, ma);
    return ma;
}


int main(){
    
    cin>>n;
    
    for(int i=0;i<n;i++){
        cin>>s[i];
    }
    int answer=0;
    for(int i=0; i<n; i++){
        for(int j=0; j<n-1; j++){
                swap(i,j); // 바꿔서 테스트
                int result = check();
                answer = max(result, answer);
                swap(i,j); // 원래대로 되돌려놓기
        }
    }
    
    for(int i=0; i<n; i++){
        for(int j=0; j<n-1; j++){
            
                swap2(j,i);
                int result = check();
                answer = max(result, answer);
                swap2(j,i);
        }
    }
    
    cout<<answer;
}

File: baekjoon/사탕게임.py
n= int(input())
arr = []
for _ in range(n):
    arr.append(list(input()))


def calculate(arr):
    for i in range(len(arr)):
        for j in range(len(arr[i])):
            print(arr[i][j], end= ' ')
        print()
    print('****************************')


def calculateY(arr, x, y):
    # [x][]
    answer=0
    tmp = 1
    for i in range(len(arr[x])-1):
        if arr[x][i] != arr[x][i+1]:
           tmp=1
        else:
            tmp+=1
            answer = max(answer, tmp)
    # [][y]
    tmp=1
    for i in range(len(arr) - 1):
        if arr[i][y] != arr[i+1][y]:
            tmp = 1
        else:
            tmp += 1
            answer = max(answer, tmp)

    # [][y+1]
    tmp = 1
    for i in range(len(arr) - 1):
        if arr[i][y+1] != arr[i + 1][y+1]:
            tmp = 1
        else:
            tmp += 1
            answer = max(answer, tmp)
    return answer

def calculateX(arr, x, y):
    # [][y]
    answer = 0
    tmp=1
    for i in range(len(arr) - 1):
        if arr[i][y] != arr[i+1][y]:
            tmp = 1
        else:
            tmp += 1
            answer = max(answer, tmp)

    # [x][]
    tmp = 1
    for i in range(len(arr[x]) - 1):
        if arr[x][i] != arr[x][i + 1]:
            tmp = 1
        else:
            tmp += 1
            answer = max(answer, tmp)

    # [x+1][]
    tmp = 1
    for i in range(len(arr[x+1]) - 1):
        if arr[x+1][i] != arr[x+1][i + 1]:
            tmp = 1
        else:
            tmp += 1
            answer = max(answer, tmp)
    return answer

answer=-1
for i in range(n-1):
    for j in range(n):
        arr[i][j], arr[i+1][j] = arr[i+1][j], arr[i][j]
        answer = max(answer, calculateX(arr, i, j))
        arr[i][j], arr[i + 1][j] = arr[i + 1][j], arr[i][j]

for i in range(n):
    for j in range(n-1):
        arr[i][j], arr[i][j+1] = arr[i][j+1], arr[i][j]
        answer = max(answer, calculateY(arr,i , j))
        arr[i][j], arr[i][j + 1] = arr[i][j + 1], arr[i][j]

print(answer)
